dartagnan/src/main/java/com/dat3m/dartagnan/wmm/Wmm.java

```diff
@@ -66,20 +66,27 @@ public Relation getRelation(String name) {
         return definition.definedRelation;
     }
 
-    public void addAlias(String name, Relation relation) {
-        addName(name, relation);
+    public Relation addName(String name, Relation relation) {
+        if (relation.named) {
+            addAlias(name, relation);
+            return relation;
+        }
+        Relation r = newRelation(name);
+        substitute(relation, r);
+        deleteRelation(relation);
```

Comment:
This code is really dangerous. By doing substitutions like this, references held by external code become outdated.
Certainly nothing that should be expected when changing the name of a relation.
I would probably do something like this:
- Relations have no name by default, likely by having a `List<String> names` of names, initially empty. The first entry of this list is the main name, other entries are aliases.
- The Wmm holds a `Set<Relation> relations` containing all relations as well as a `Map<String, Relation> aliasMap` for the names/aliases.
- When calling `addName/addAlias`, the map is checked for duplicates. If none, the map is updated and the name is appended to `rel.names`. The first added alias is automatically the main name of the relation.
- Maybe there is a `setName` function to set the main name (either inserting the name at the front, or moving an already existing alias to the front). We have no use-case for this yet though.
- I also don't think we need to be able to do lookups of relations by term, so we can get rid of those and only store true names/aliases. Search by term can be done by iterating over the set of relations, if necessary.

I think you won't get any cleaner solution than the above. It supports exactly what is needed (arbitrary number of names per relation) without any hidden traps and moderate state management.

Reply:
I was glad that I found a way to describe a relation's string representation that is both final (and therefore robust against restructuring) and human-readable.  Since the ability to rename was not required and the names of relations are known to `VisitorBase` before the creation of the relation, there should be no reason not to give it a name setter, yet.
I agree with you, concerning the substitution feature.  The only current usage is unnecessary.  I will remove it.
The term mapping implements a feature.  Duplicate expressions like `co | fr` in `tso.cat` are recognized and unified into one relation.  I did not plan to remove this, but I could do it, as cases like that will be rare and easily fixable in the input.  In the above case, both duplicates are removed during simplification.
As for the handling of aliases, after all those fixes were applied, I saw no obstacle to add that, as well.

Reply:
>I was glad that I found a way to describe a relation's string representation that is both final (and therefore robust against restructuring) and human-readable. Since the ability to rename was not required and the names of relations are known to VisitorBase before their creation, there should be no reason to let them keep a name setter.

Technically, relations can be unnamed (which is also useful). By forcing default names onto those relations, you end up having to do special case reasoning to distinguish between named and unnamed relations, which feels awkward.
While this allows you to have a final string representation (using the dummy names), I don't see why this would be beneficial (it's also less human-readable IMO). After all, our whole Wmm structure is not immutable and is also not intended to be.
The new implementation of `Relation` also does not check equality based on naming, or computes its hashcode from it.
This gives it a lot less reason to have a final name on construction, particularly a non-null one.
Lastly, I think there are cases where the name of a relation is unknown during its first construction:
```
let com = rf | co | rf^-1;co // used fr, but without its name
let fr = rf^-1;co // Only here, the relation is named for the first time
```
I guess a processing step after the parsing could also eliminate these duplicates, but it might be easiest to do this already during the parsing (like it is currently done, IIRC). With fixed initial names, you would need to reconstruct the whole Wmm (which is not too bad in terms of performance, but might feel too clunky in practice).

>The term mapping implements a feature. Duplicate expressions like co | fr in tso.cat are recognized and unified into one relation. I did not plan to remove this, but I could do it, as cases like that will be rare and easily fixable in the input. In the above case, both duplicates are removed during simplification.

Ok, I see that in this case, having fixed names and fixed terms is a useful feature. We wouldn't want to have structurally equivalent duplicates like multiple `(co | fr)` in our Wmm. But I don't think this feature necessitates a term mapping in the Wmm. The parser itself could likely maintain a `Term -> Relation` map, or query the Wmm repeatedly for terms it is about to construct.

Overall, I think there are only two reasonable approaches:
- Make the whole Wmm, its relations, and its axioms all immutable. In this case, the Wmm is seen as a big syntactical expression and you need a Transformer/Rewriter to do any changes. This has nice properties but might feel overly clunky in practice (recursiveness + immutability is also a big problem).
- Make the Wmm, its relations and its axiomsm mutable. This is what general graph structures usually are. This makes (simple) transformations easy to implement but requires careful state management to avoid inconsistent data. The state management should therefore be kept to a minimum. It allows for more freedom but also for more ways to break the state. In practice, however, I believe this to be a minor issue because we follow general principles like
-- not relying on cached state during transformation/processing phases
-- not keeping state from the transformation/processing phases
-- not mutating the data during the analysis phase

Our old implementation did follow the first approach of immutability (not full immutablity though).
I felt that working with this was a pain when it came to doing any sorts of transformations of even analyzing it properly (dummy recursive relations that were not returned by `.getRelations`???).
So our new approach should remedy this issue, even if it introduces some more ways to break data consistency.

Reply:
I forgot to add: the default names `_x` are just numbers in disguise, no? Maybe you are better off giving all Relations an `id` rather than a name by default.

Reply:
If there was any other use for that `id` other than identifying the instance and being transformed into a string.  But this way relation instances are unique anyway; there is no need to define equivalence classes by id.  Therefore I do the string transformation once instead of each time a name is fetched for a edge or clock variable.

I think I should replace the `_` prefix with something like `#`, to avoid name collision with input-defined relations.

Reply:
I think the underscore is totally fine. But I don't mind using any other special symbol instead.

Reply:
Sounds good. 
The thing about naming a relation the same as a term is one of those "issues" that are practically non-existant. I wouldn't bother with that at all. Writing defensive code against that makes everything more complicated for virtually zero benefit.